File: 264.py
# 注意：这份代码并没有ac（笑
class Solution:
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        lis = [0]*n
        num = [2, 3, 5]
        lis[:5] = [1, 2, 3, 4, 5]
        for i in range(5, n):
            m = 2*lis[i-1]
            for p in lis[:i][::-1]:
                if p*5 <= lis[i-1]:
                    break
                for j in num:
                    k = j*p
                    if k >= m:
                        break
                    if k > lis[i-1]:
                        m = k
            lis[i] = m
        return lis[n-1]

File: 264.cpp
// cpp过了python没过...
class Solution {
public:
    int nthUglyNumber(int n) {
        long lis[n], m, su;
        int num[3] = { 2, 3, 5 };
        int idx;
        lis[0] = 1;
        for (int i = 1; i < n; i++) {
            m = 2 * lis[i-1];
            for (int j = 0; j < i; j++) {
                idx = i-j-1;
                if (5 * lis[idx] <= lis[i-1]) break;
                for (int k = 0; k < 3; k++) {
                    su = num[k] * lis[idx];
                    if (su >= m) break;
                    if (su > lis[i-1]) m = su;
                }
            }
            lis[i] = m;
        }
        return lis[n-1];
    }
};